cpp/src/tsvd/tsvd_mg.cu

```diff
@@ -0,0 +1,436 @@
+ * @input param verbose
+ */
+template <typename T>
+void inverse_transform_impl(cumlHandle &handle,
+                            Matrix::RankSizePair **rank_sizes, size_t n_parts,
+                            Matrix::Data<T> **trans_input, T *components,
+                            Matrix::Data<T> **input, paramsTSVD prms,
+                            bool verbose) {
+  int rank = handle.getImpl().getCommunicator().getRank();
+
+  std::vector<Matrix::RankSizePair *> ranksAndSizes(rank_sizes,
+                                                    rank_sizes + n_parts);
+  Matrix::PartDescriptor trans_desc(prms.n_rows, prms.n_components,
+                                    ranksAndSizes, rank);
+  std::vector<Matrix::Data<T> *> trans_data(trans_input, trans_input + n_parts);
+
+  std::vector<Matrix::Data<T> *> input_data(input, input + n_parts);
+
+  // TODO: These streams should come from cumlHandle
+  int n_streams = n_parts;
+  cudaStream_t streams[n_streams];
+  for (int i = 0; i < n_streams; i++) {
+    CUDA_CHECK(cudaStreamCreate(&streams[i]));
+  }
+
+  inverse_transform_impl(handle, trans_data, trans_desc, components, input_data,
+                         prms, streams, n_streams, verbose);
+
+  for (int i = 0; i < n_streams; i++) {
+    CUDA_CHECK(cudaStreamSynchronize(streams[i]));
+  }
+
+  for (int i = 0; i < n_streams; i++) {
+    CUDA_CHECK(cudaStreamDestroy(streams[i]));
+  }
+}
+
+/**
+ * @brief performs MNMG fit and transform operation for the tsvd.
+ * @input param handle: the internal cuml handle object
+ * @input param rank_sizes: includes all the partition size information for the rank
+ * @input param n_parts: number of partitions
+ * @input param input: input data
+ * @output param trans_input: transformed input data
+ * @output param components: principal components of the input data
+ * @output param explained_var: explained var
+ * @output param explained_var_ratio: the explained var ratio
+ * @output param singular_vals: singular values of the data
+ * @input param prms: data structure that includes all the parameters from input size to algorithm
+ * @input param verbose
+ */
+template <typename T>
+void fit_transform_impl(cumlHandle &handle,
+                        std::vector<Matrix::Data<T> *> &input_data,
+                        Matrix::PartDescriptor &input_desc,
+                        std::vector<Matrix::Data<T> *> &trans_data,
+                        Matrix::PartDescriptor &trans_desc, T *components,
+                        T *explained_var, T *explained_var_ratio,
+                        T *singular_vals, paramsTSVD prms, bool verbose) {
+  int rank = handle.getImpl().getCommunicator().getRank();
+
+  // TODO: These streams should come from cumlHandle
+  int n_streams = input_desc.blocksOwnedBy(rank).size();
+  ;
+  cudaStream_t streams[n_streams];
+  for (int i = 0; i < n_streams; i++) {
+    CUDA_CHECK(cudaStreamCreate(&streams[i]));
```

Comment:
I know we've talked at length about this, however now that we're passing in the cuml handle, we should add this to the Github issue to pull these streams out of it, rather than creating them.

Reply:
We should reference the issue here